### PR TITLE
feat(core): Add local symlink mounts to LocalSandbox (COR-495, COR-554)

### PR DESCRIPTION
**Pull request by NikAiyer**
## Summary

- Add symlink-based mount support to `LocalSandbox` for local filesystem mounts
- When a `LocalFilesystem` is mounted, a symlink is created from `<workingDirectory>/<mountPath>` to the `basePath`, allowing shell commands to access mounted data directly
- Add shared mount infrastructure (forward-compatible `MountResult.unavailable`, `MountState: 'unavailable'`, `MountToolNotFoundError`) that PR #13415 (FUSE mounts) builds on
- Add `contained: false` warning for mount configs (COR-554)
- Add mount path resolution in workspace instructions

### Files changed

| File | Change |
|------|--------|
| `local-sandbox.ts` | `mount()` with `case 'local'` (symlink), `unmount()`, marker files, path validation, isolation allowlist |
| `local-filesystem.ts` | `getMountConfig()` returning `{ type: 'local', basePath }`, `contained` getter |
| `mount-manager.ts` | `unavailable` state handling, `MountToolNotFoundError` catch |
| `sandbox.ts` | Remove `workingDirectory` from interface (now via `instanceof LocalSandbox`) |
| `types.ts` | `MountState: 'unavailable'`, `MountToolNotFoundError` class |
| `mount.ts` | `MountResult.unavailable` field |
| `workspace.ts` | `contained: false` warning, mount path resolution with `workingDir`, `getPathContext` via `instanceof` |
| Tests | 17 symlink mount tests, 3 contained:false tests, 2 LocalFilesystem tests |

## Stacking

This is PR 1 of 2. PR #13415 (`feat-local-sandbox-mounts`) adds FUSE mount support (S3/GCS) on top of this branch.

## Test plan

- [x] 122 LocalSandbox tests pass (including 17 new mount operation tests)
- [x] 195 Workspace + LocalFilesystem tests pass (including 3 new contained:false tests + 2 new getMountConfig tests)
- [x] TypeScript type check passes
- [x] ESLint passes
- [ ] CI passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Local symlink mount support so sandboxed commands can access host paths with automatic mount/unmount lifecycle.
  * Mount management surfaced via new public exports for better integration and diagnostics.

* **Bug Fixes**
  * Safer mount path validation, improved host-path resolution, and cleanup on stop/destroy.
  * Mounts can now report an "unavailable" state with clearer messaging when required tooling is missing.
  * Warns when a local filesystem mount is configured as non-contained; prefer contained or explicit allowed paths.

* **Documentation**
  * Workspace instructions show resolved mount locations and relevant warnings.

* **Tests**
  * Expanded mount-related coverage and edge-case validations.

### CODE REVIEW

File: .changeset/whole-grapes-lay.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+const workspace = new Workspace({
+  mounts: {
+    '/data': new LocalFilesystem({ basePath: '/path/to/data' }),
+  },
+  sandbox: new LocalSandbox({ workingDirectory: './workspace' }),
+});
+
+await workspace.init();
+// Sandboxed commands can access the mount path via symlink
+await workspace.sandbox.executeCommand('ls', ['data']);
```
Comment (CalebBarnes) [inline .changeset/whole-grapes-lay.md]:
```suggestion
await workspace.sandbox.executeCommand('ls data');
```

File: packages/core/src/workspace/sandbox/local-sandbox.ts [hunk 2]
```diff
@@ -315,6 +376,347 @@ export class LocalSandbox extends MastraSandbox {
...
+          await fs.rmdir(hostPath);
+          await fs.symlink(localConfig.basePath, hostPath);
+          this.logger.debug(`[LocalSandbox] Symlinked local mount ${hostPath} → ${localConfig.basePath}`);
+          break;
+        }
+        default:
+          try {
+            await fs.rmdir(hostPath);
+          } catch {
+            // best-effort cleanup
```
Comment (CalebBarnes) [inline packages/core/src/workspace/sandbox/local-sandbox.ts]: Shouldn't we just check if its config.type === 'local' at the very start before doing any of the work?

File: packages/core/src/workspace/sandbox/local-sandbox.ts [hunk 3]
```diff
@@ -315,6 +376,347 @@ export class LocalSandbox extends MastraSandbox {
...
+    } catch {
+      // Ignore if doesn't exist
+    }
+
+    // Remove mount point (symlink or empty directory)
+    try {
+      if (isSymlink) {
+        await fs.unlink(hostPath);
+      } else {
+        await fs.rmdir(hostPath);
```
Comment (CalebBarnes) [inline packages/core/src/workspace/sandbox/local-sandbox.ts]: Since we only symlink for local mount, avoid rmdir to prevent unintentional deletions.

File: packages/core/src/workspace/sandbox/local-sandbox.ts [hunk 4]
```diff
@@ -315,6 +376,347 @@ export class LocalSandbox extends MastraSandbox {
...
+        return { success: false, mountPath, error };
+      }
+      // ENOENT: dir doesn't exist yet (mkdir below creates it)
+      // Other: proceed; mkdir will surface the real error
+    }
+
+    // Create mount directory under working directory
+    try {
+      this.logger.debug(`[LocalSandbox] Creating mount directory at ${hostPath}...`);
+      await fs.mkdir(hostPath, { recursive: true });
```
Comment (CalebBarnes) [inline packages/core/src/workspace/sandbox/local-sandbox.ts]: Is it necessary to create the dir? we just rmdir it later and then create a symlink here.

File: packages/core/src/workspace/sandbox/local-sandbox.ts [hunk 5]
```diff
@@ -315,6 +376,347 @@ export class LocalSandbox extends MastraSandbox {
...
+      return { success: false, mountPath, error: String(mkdirError) };
+    }
+
+    try {
+      switch (config.type) {
+        case 'local': {
+          // Local filesystem — create a symlink from hostPath to the basePath
+          const localConfig = config as { type: 'local'; basePath: string };
+          // Remove the empty directory created above — symlink replaces it
+          await fs.rmdir(hostPath);
```
Comment (CalebBarnes) [inline packages/core/src/workspace/sandbox/local-sandbox.ts]: Do we actually need to create it just to delete it? Is it possible to just make symlink instead?

**Review by CalebBarnes — APPROVED** [review]
